utils/simulator/simulator.py

```diff
@@ -0,0 +1,198 @@
+
+        # connect over tcp to scriptio console
+        process_start_timestamp = asyncio.get_event_loop().time()
+        while asyncio.get_event_loop().time() - process_start_timestamp < 10.0:
+            try:
+                scriptio_reader, scriptio_writer = await asyncio.open_connection(
+                    'localhost', scriptio_port)
+            except OSError:
+                continue
+            else:
+                break
+
+        # even in no-echo mode, the scriptio console currently prints one line at the beginning
+        # so lets read it
+        await scriptio_reader.readline()
+
+        # start parsing stdout/logs
+        logs = Publisher()
+
+        async def parse_simulator_stdout():
+            while process.stdout and not process.stdout.at_eof():
+                line = (await
+                        process.stdout.readline()).decode('utf-8').strip()
+                logger.info('%s', line)
+                await logs.publish(line)
+
+        logs_task = asyncio.ensure_future(parse_simulator_stdout())
+
+        # yield the simulator to the callee
+        try:
+            yield Simulator(process=process,
+                            machine=machine,
+                            tmpdir=tmpdir,
+                            logs=logs,
+                            scriptio_reader=scriptio_reader,
+                            scriptio_writer=scriptio_writer,
+                            http_proxy_port=http_proxy_port)
+        finally:
+            scriptio_writer.close()  # type: ignore
+            await scriptio_writer.wait_closed()  # type: ignore
+            process.terminate()
+            logs_task.cancel()
+            await process.communicate()
+
+    @staticmethod
+    def default_simulator_path() -> Optional[Path]:
+        mini404_dep_path = get_dependency_directory('mini404')
+        simulator_path = mini404_dep_path / 'qemu-system-buddy'
+        if simulator_path.exists():
+            return simulator_path
+        else:
+            return None
+
+    async def command(self, command: str, readline=False):
+        async def issue_command():
+            assert self.scriptio_writer, 'scriptio socket isn\'t connected'
+            self.scriptio_writer.write(command.encode('utf-8') + b'\n')
+
+        async def wait_for_script_finished_line():
+            async for line in self.logs:
+                if line.strip() == 'ScriptHost: Script FINISHED':
+                    break
+
+        await asyncio.gather(issue_command(), wait_for_script_finished_line())
+
+        if readline:
+            line = await self.scriptio_reader.readline()
+            return line.decode('utf-8').strip()
+
+    #
+    # encoder primitives
+    #
+
+    async def encoder_click(self):
+        await self.command('encoder-input::Push()')
+
+    async def encoder_push(self):
+        raise NotImplementedError()
+
+    async def encoder_release(self):
+        raise NotImplementedError()
```

Comment:
Are these two harder to implement, or is it worth just throwing them in while we are at it?

Reply:
Yes, there is no support in the simulator yet. My intention is to have them here to clearly differentiate push/release and click.